[AVR] No cli for SPWRITE on XMEGA

tomtor opened the pull request:
From the XMEGA series manual:
_To prevent corruption when updating the stack pointer from software,
a write to SPL will automatically disable interrupts
for up to four instructions or until the next I/O memory write._

This will save 6! instructions in every function with an FP on ALL Xmega devices and it
is compatible with gcc code generation for all those devices. It is also not very modern, because the first devices are from 2013. So this is really significant and it is a small change.

@benshi001 Note that I could not find any existing (SPWRITE) unit tests to which this variant could be added.

tomtor committed:
[AVR] No cli for SPWRITE on XMEGA

tomtor committed:
formatting

benshi001 commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+        .addReg(SrcLoReg, getKillRegState(SrcIsKill))
+        .setMIFlags(Flags);
+
+    buildMI(MBB, MBBI, AVR::OUTARr)
```
It would be better to check if SPH is available here, for example
```c++
if (getIORegSPH() >= 0)
  ....
```

tomtor replied:
Original code

benshi001 commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+  // To prevent corruption when updating the stack pointer from software,
+  // a write to SPL will automatically disable interrupts
+  // for up to four instructions or until the next I/O memory write.
+  if (STI.getELFArch() >= 102) { // An XMEGA device
```
Is it applicable to all upper devices? should xmega3 be excluded ?

tomtor replied:
It should apply to all

benshi001 commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+  // To prevent corruption when updating the stack pointer from software,
+  // a write to SPL will automatically disable interrupts
+  // for up to four instructions or until the next I/O memory write.
+  if (STI.getELFArch() >= 102) { // An XMEGA device
```
Is it applicable to all upper devices? should xmega3 be excluded ?

benshi001 commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+        .addImm(0x3d)
+        .addReg(SrcLoReg, getKillRegState(SrcIsKill))
+        .setMIFlags(Flags);
+  }
```
A unit test is needed for this change. Please see other tests for pseudo instructions at `llvm/test/CodeGen/AVR/pseudo`.

tomtor replied:
None of those vary architecture?

benshi001 replied:
There are tests in this directory show difference behavior for different architectures, for example OUTWARr.mir

benshi001 commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+        .addImm(0x3d)
+        .addReg(SrcLoReg, getKillRegState(SrcIsKill))
+        .setMIFlags(Flags);
+  }
```
Please do not use `git push -f ` when updating a PR, I need to trace if all of my concerns are fixed.

benshi001 replied:
just use `push`, more than one commits will be squashed.

tomtor replied:
I never used `push -f` in this PR, but did indeed in earlier ones.

tomtor commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
-      .addImm(0x3d)
-      .addReg(SrcLoReg, getKillRegState(SrcIsKill))
-      .setMIFlags(Flags);
+  // From the XMEGA series manual:
```
https://ww1.microchip.com/downloads/en/DeviceDoc/Atmel-8331-8-and-16-bit-AVR-Microcontroller-XMEGA-AU_Manual.pdf page 17 SPL description.

It is in every chip full manual in the section describing SPL

benshi001 replied:
Thanks. I will learn it later.

tomtor commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+  if (STI.getELFArch() >= 102) { // An XMEGA device
+
+    buildMI(MBB, MBBI, AVR::OUTARr)
+        .addImm(0x3d)
```
Ok, i saw that test, but then it is not clear that this code applies to every newer avr device.

benshi001 replied:
You are appreciated to help fix that issue.

tomtor commented:
Regarding gcc output, compile any C fragment with gcc with eg -mmcu=attiny85 and -mmcu=attiny1614 and compare the start and end of the different -S output.

tomtor commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+        .addReg(SrcHiReg, getKillRegState(SrcIsKill))
+        .setMIFlags(Flags);
+
+    buildMI(MBB, MBBI, AVR::OUTARr)
```
Ah, from SEI in the manual:

Sets the Global Interrupt Enable (I) bit in SREG (Status Register). The instruction following SEI will be executed
before any pending interrupts.

benshi001 replied:
Thanks. I understand it now.

tomtor commented on llvm/lib/Target/AVR/AVRExpandPseudoInsts.cpp:
```diff
@@ -2531,27 +2531,46 @@ bool AVRExpandPseudo::expand<AVR::SPWRITE>(Block &MBB, BlockIt MBBI) {
+
+  } else { // Disable interrupts for older devices (3 extra instructions)
+
+    buildMI(MBB, MBBI, AVR::INRdA)
```
SPL and SPH order is also reversed, so I do not think this is possible, clear to read?

benshi001 replied:
Sorry for the delay. But I am working on llvm in my spare time. 

For this PR, keep the original structure is OK, due to the byte order issue.

tomtor committed:
reorder

tomtor committed:
no constants for SP location

tomtor committed:
Test and optimize for SPL only device

tomtor committed:
Formatting for SPL only test

tomtor committed:
Another SPL only test

tomtor committed:
Fix another SPL only test

tomtor committed:
Revert "Fix another SPL only test"

This reverts commit 2a3a9507704efc64b9e6d152b1da4aca6aaa15ad.

tomtor committed:
Revert "Another SPL only test"

This reverts commit 82c45cbe49059ff87f373b3b255303586f1b4d88.

tomtor committed:
Revert "Formatting for SPL only test"

This reverts commit 28b238abb31eda84a3f8892d1d5e0f78cb128d79.

tomtor committed:
Revert "Test and optimize for SPL only device"

This reverts commit 2e7629e65a1333598c63a8cd8a0b753a1342dd42.

tomtor committed:
Revert "no constants for SP location"

This reverts commit f80f924bf212e3938f2f552aedbbe3532472a6c9.

tomtor committed:
Revert "reorder"

This reverts commit 24afcd7c514fa153f9fe64e1f55c48e3ede49407.

benshi001 reviewed (commented):
.

benshi001 commented:
@tomtor Sorry for my late response. I also work on llvm in my spare time without any payment. I can not serve reviewing works in time. For that reason, my two core points are

1. Stability is much more important than optimization, especially bug fixes. I would rather sub optimized code is generated than broken code. So I quite concern if optimization for newer devices will harm older devices.
2. Fix defects against avr-gcc-7.3 tool chain, arduino still uses gcc-7.3 as its main tool chain.

tomtor committed:
use RegSPL/H instead of inline constants

benshi001 commented:
For the unit test to be supplemented, a contrast test between `avr6` and `avrxemaga3` is enough.

tomtor committed:
Add tests and check to prevent write -1

tomtor committed:
Add tests and check to prevent write -1

tomtor committed:
Format Add tests and check to prevent write -1

tomtor commented:
> For the unit test to be supplemented, a contrast test between `avr6` and `avrxemaga3` is enough.

I added avr128db28 and attiny1614. Note that we MUST test for SPH to prevent writing wrong assembly with -1 as destination, so I added that single test.

So we have 3 tests, a classic (atmega328), a tiny without SPH and some modern xmegas/tiny/avr

tomtor committed:
remove empty line
